File: AddTwoNumbers.java

class ListNode {
	int val;
	ListNode next;

	ListNode() {
	}

	ListNode(int val) {
		this.val = val;
	}

	ListNode(int val, ListNode next) {
		this.val = val;
		this.next = next;
	}

}

public class AddTwoNumbers {

	public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {

		// listnode for output
		ListNode list = new ListNode(0);
		ListNode output = list;
		// digit to carry over
		int carryOver = 0;

		while (l1 != null || l2 != null) {

			int sum = carryOver;
			if (l1 != null) {
				sum += l1.val;
				l1 = l1.next;
			}

			if (l2 != null) {
				sum += l2.val;
				l2 = l2.next;
			}

			if (sum > 9) {
				carryOver = 1;
				sum = sum - 10;
			} else {
				carryOver = 0;
			}

			ListNode l = new ListNode(sum);
			list.next = l;
			list = list.next;

		}

		if (carryOver > 0) {
			ListNode l = new ListNode(carryOver);
			list.next = l;
		}

		return output.next;
	}

	public static void main(String[] args) {
		ListNode l3 = new ListNode(3);
		ListNode l2 = new ListNode(4, l3);
		ListNode l1 = new ListNode(2, l2);

		ListNode l6 = new ListNode(4);
		ListNode l5 = new ListNode(6, l6);
		ListNode l4 = new ListNode(5, l5);

		ListNode output = addTwoNumbers(l1, l4);
		while (output != null) {
			System.out.println(output.val);
			output = output.next;
		}
	}
}

File: AddTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    // use two pointers to iterate through both lists
    let ptr1 = l1;
    let ptr2 = l2;

    // use a carry variable to keep track of the carry
    let carry = 0;

    // use a dummy node to keep track of the head of the output list
    let output = new ListNode();
    let ptr3 = output;

    // while there are still numbers to add
    while (ptr1 || ptr2) {
        // initialize the sum to be the carry from before
        let sum = carry;
        // if there is a number in the first list, add it to the sum and move the pointer
        if (ptr1) {
            sum += ptr1.val;
            ptr1 = ptr1.next;
        }
        // if there is a number in the second list, add it to the sum and move the pointer
        if (ptr2) {
            sum += ptr2.val;
            ptr2 = ptr2.next;
        }
        // if the sum is greater than 10, set the carry to be 1 and subtract 10 from the sum
        if (sum >= 10) {
            carry = 1;
            sum -= 10;
        }
        // else, set the carry to be 0
        else {
            carry = 0;
        }
        // move the pointer for the output list and set the value to be the sum
        ptr3.next = new ListNode(sum);
        ptr3 = ptr3.next;
    }
    
    // if there is still a carry after iterating, add it to the end of the output list
    if (carry) {
        ptr3.next = new ListNode(carry);
    }

    // return the head of the output list
    return output.next;
};